File: labs/lab-app-egress-audit/gateway/capture.py
"""mitmproxy addon: record the three egress visibility layers to JSONL.

- tls_clienthello: the SNI of every TLS connection, decryptable or not.
- request: fully decrypted HTTP for connections whose cert the app trusted.
- tls_failed_client: the app rejected our forged cert (certificate pinning) —
  the connection stays opaque, but the SNI is still a finding.
"""

import json
import os
import threading

LOG = os.environ.get("CAPTURE_LOG", "/capture/capture.jsonl")
_lock = threading.Lock()


def _write(record):
    with _lock:
        with open(LOG, "a") as fh:
            fh.write(json.dumps(record) + "\n")
            fh.flush()


class Capture:
    def __init__(self):
        self._sni_by_client = {}

    def tls_clienthello(self, data):
        sni = data.client_hello.sni
        self._sni_by_client[data.context.client.id] = sni
        _write({"event": "clienthello", "sni": sni})

    def request(self, flow):
        _write(
            {
                "event": "request",
                "host": flow.request.pretty_host,
                "method": flow.request.method,
                "path": flow.request.path,
                "body": flow.request.get_text(strict=False),
            }
        )

    def tls_failed_client(self, data):
        sni = self._sni_by_client.get(data.context.client.id)
        _write({"event": "tls_failed_client", "sni": sni})


addons = [Capture()]

File: labs/lab-app-egress-audit/analyzer/src/capture.ts
export type HostEvidence = {
  decrypted: boolean;
  sniSeen: boolean;
  tlsFailed: boolean;
  payload?: string;
};

type Row =
  | { event: "clienthello"; sni: string | null }
  | { event: "request"; host: string; method: string; path: string; body: string | null }
  | { event: "tls_failed_client"; sni: string | null };

function ensure(map: Map<string, HostEvidence>, fqdn: string): HostEvidence {
  let ev = map.get(fqdn);
  if (!ev) {
    ev = { decrypted: false, sniSeen: false, tlsFailed: false };
    map.set(fqdn, ev);
  }
  return ev;
}

export function parseCapture(jsonl: string): Map<string, HostEvidence> {
  const map = new Map<string, HostEvidence>();
  for (const line of jsonl.split("\n")) {
    if (!line.trim()) continue;
    const row = JSON.parse(line) as Row;
    if (row.event === "clienthello" && row.sni) {
      ensure(map, row.sni).sniSeen = true;
    } else if (row.event === "request") {
      const ev = ensure(map, row.host);
      ev.decrypted = true;
      ev.sniSeen = true;
      if (row.body) ev.payload = row.body;
    } else if (row.event === "tls_failed_client" && row.sni) {
      ensure(map, row.sni).tlsFailed = true;
    }
  }
  return map;
}
